pravega/pravega
# Issue 5454: Fixing Segment Store flaky tests and other bugs

**Change log description**  
1. Fixing various unit tests in the Segment Store package(s) that caused builds to randomly fail.
2. Fixed a bug in `BlockingDrainingQueue` that would have resulted in data loss if a TimeoutException preempted the completion of a timeout-eligible take future (#5516).
3. Fixed a bug in `OperationProcessor` error handler to initiate the service shutdown in an async manner (to avoid deadlocks)
4. Fixed a bug in `OperationProcessor` to "unaccount" for the cache effects of operations that are in flight but have failed to be committed due to the service shutting down (byte accounting leak).
5. Fixed a bug in `StreamSegmentReadIndex` that would have caused a ReadResult to hang if certain concurrency conditions occur (merging a source segment into a parent one at the same time as either issuing the ReadResultEntry or completing it).
6. Fixed a bug in `RedirectedReadIndexEntry` that would throw an `UnsupportedOperationException` when attempting to close. This is invoked when a `ReadResult` is cancelled/cleaned up (so if it's sitting on one of those entries, the operation would fail).
7. Fixed a bug in `StreamSegmentContainer` that would incorrectly report as "Failed" if it is immediately shut down after startup (i.e., if the secondary services haven't yet been started).

**Purpose of the change**  
Fixes #5516.
Fixes #5454.
Fixes #5499.
Fixes #5492.
Fixes #5490.
Fixes #5487.
Fixes #5484.
Fixes #5500.
Fixes #5495.
Fixes #5491.
Fixes #5485.
Fixes #5505.
Fixes #5480.

**What the code does**  
See description above.

**How to verify it**  
Build must pass both on Travis and GitHub actions. This PR does not address all sporadic failures, but should be a good step towards a more stable build.
